### ISSUE TITLE
Logging: reopen logs after reload

### ISSUE THREAD

**Issue by motiejus** [caddyserver/caddy#5316]
Summary: when Caddy is reloaded, it should re-open its log files. If the file does not exist, create one.

Use case: I would like logrotate to rotate the log files. This is the convention with logrotate services (copied from https://github.com/caddyserver/caddy/issues/329#issuecomment-223819543). When time to rotate the configs come:

1. move the log file (e.g. `mv access.log access.log.1`).
2. reload the process: `systemctl reload caddy`.
3. let the process close the log file. There is no race condition, because the process kept writing to the file with the new name until now.
4. let the process re-open the log file with old (original) name. *This creates a new file.*
5. do something with renamed log file (compress)

However, step (4) does not work: Caddy does not re-open the file.

### Steps to reproduce

**Caddyfile**
```
{
    http_port 8088
    https_port 8084
}

http://localhost:8088 {
    file_server 

    log {
        output file access.log
    }
}
```

1. Run the server: `caddy run --config Caddyfile`
2. Issue a request: `curl http://localhost:8088`
3. Observe there is a log line in `access.log`.
4. `mv access.log access.log.1`
6. caddy reload --config Caddyfile`
7. Issue a 2nd request: `curl http://localhost:8088`
8. Observe that the logs of the second request are in `access.log.1`, and `access.log` was not created.

### What did I expect instead?

At step (6) a new file `access.log` was created and it has the mentioned log line.

**Comment by francislavoie** [comment]
Reloading the config when there's no actual changes to the config is a no-op, unless you use the `--force` flag.

Might that be what's tripping you up?

**Comment by motiejus** [comment]
> Reloading the config when there's no actual changes to the config is a no-op, unless you use the `--force` flag.
> 
> Might that be what's tripping you up?

I observed this in production. Updating the description.
> 
> FYI, Caddy has its own log rolling, see https://caddyserver.com/docs/caddyfile/directives/log#file (but maybe that's what you meant you're using).

No, I am by choice using logrotate.

**Comment by Forza-tng** [comment]
> I am by choice using logrotate.

You can use the `copytruncate` option with logrotate. This is maybe not very efficient on large (as Caddy logs can be) files.

Perhaps a new `caddy rotate logs` option could be created, which will either force a log rotate  (Caddy's internal logrotate) or re-opens log files if `roll_disabled` is set.

**Comment by mholt** [comment]
Does the logger even return an error on `Write()` if the file is removed out from under it? IIRC, no error is returned -- I may be wrong though. If I'm right, however, then I'm really not sure how we can do it, because I think a failed write is the only way we'd know that the file is missing.

**Comment by mholt** [comment]
> Anyhow, the convention is SIGHUP.

Is this reliable though? And what does it tell us exactly? There's sometimes more than one log output. I guess we just reset all of them? Does the SIGHUP happen after the file is replaced out from under the feet of the writer? Or does it happen before? If after, how many log lines can get lost before we can receive and process the signal?

I have so many questions!

Stat'ing on each write is too expensive, I don't think we can afford that.

**Comment by Forza-tng** [comment]
> Does the SIGHUP happen after the file is replaced out from under the feet of the writer?

The normal case, AFAIK, is that when the original log file is renamed (mv), the writes continue to the new name because the fd and inode is unchanged. A new file is created in its place and when sighup is sent, the process opens the new file and continues writing there. So no lost writes should happen.

**Comment by mohammed90** [comment]
As mentioned earlier, mv-ing a file does not change its inode or fd, so Caddy will keep writing to the same inode, which in th example of OP is `access.log.1`. It will not and cannot switch to the newly created empty file, i.e. `access.log`, unless Caddy closes the old fd/inode and opens the new one. As @francislavoie said, it may be wasteful to close/reopen the log file on every reload, and I perceive it may be subject to abuse the system.

> > I am by choice using logrotate.
> 
> You can use the `copytruncate` option with logrotate. This is maybe not very efficient on large (as Caddy logs can be) files.

Without knowing much about your environment and based on the limited information given, this appears to be the best option for your use case. Will it not work for you?

The other option may be to use the [github.com/fsnotify/fsnotify](github.com/fsnotify/fsnotify) package to detect file changes. However, same with closing/reopening the file, this introduces an opportunity of abuse/misuse. Caddy may be spammed with a flood of (bogus) events causing it to constantly close/open the file, potentially blocking log writes as the logger awaits a proper and usable fd.

**Comment by motiejus** [comment]
> As mentioned earlier, mv-ing a file does not change its inode or fd, so Caddy will keep writing to the same inode, which in th example of OP is `access.log.1`. It will not and cannot switch to the newly created empty file, i.e. `access.log`, unless Caddy closes the old fd/inode and opens the new one. As @francislavoie said, it may be wasteful to close/reopen the log file on every reload, and I perceive it may be subject to abuse the system.

My machine can open and close a file about 200k times per second. And that's a mediocre laptop on an encrypted btrfs filesystem (i.e. overhead). Just tried this on a server, got almost 500k/s. We should be reloading the configuration hundreds of times per second for this to matter. I guess we aren't?

<details><summary>main.go</summary>
package main

import (
	"fmt"
	"os"
	"time"
)

const nfiles = 1_000_000

func main() {
	start := time.Now()
	for i := 0; i < nfiles; i++ {
		f, err := os.OpenFile("tempfile.txt", os.O_CREATE, 0644)
		if err != nil {
			fmt.Printf("error opening file: %v\n", err)
			os.Exit(1)
		}
		f.Close()
	}
	duration := time.Now().Sub(start)

	if err := os.Remove("tempfile.txt"); err != nil {
		fmt.Printf("error deleting file: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("It took %.2f seconds to create %d files.\n", duration.Seconds(), nfiles)
	fmt.Printf("That's %.f per second.\n", nfiles/duration.Seconds())

}
</details>

My laptop:
```
$ go run main.go
It took 4.55 seconds to create 1000000 files.
That's 219605 per second.
```
Some server in GCP I have access to:
```
It took 2.07 seconds to create 1000000 files.
That's 483532 per second.
```

> > > I am by choice using logrotate.
> > 
> > 
> > You can use the `copytruncate` option with logrotate. This is maybe not very efficient on large (as Caddy logs can be) files.

Like mentioned in the documentation of copytruncate, it does have a race condition which may cause it to lose data. For me it is more acceptable to restart caddy and lose a bit of traffic.

> Without knowing much about your environment and based on the limited information given, this appears to be the best option for your use case. Will it not work for you?

I have an "ecosystem" of log files and know how to deal with them. :) Here is a logrotate configuration for a web server that I wrote (which I am planning to replace caddy with):

```
/var/log/private/jjweb/access.log {
    rotate 60
    daily
    dateext
    dateyesterday
    compress
    compresscmd /usr/bin/zstd
    compressext .zst
    compressoptions --long -19
    uncompresscmd /usr/bin/unzstd
    postrotate
        /usr/bin/systemctl reload jjweb
    endscript
}
```

Logrotate does not natively support those options. The most important for me are `compresscmd`, `dateext` and `dateyesterday`.

> The other option may be to use the [github.com/fsnotify/fsnotify](github.com/fsnotify/fsnotify) package to detect file changes (keeping in mind the complexity of robust handling of the events: [fsnotify/fsnotify#372](https://github.com/fsnotify/fsnotify/issues/372)). However, same with closing/reopening the file, this introduces an opportunity of abuse/misuse. Caddy may be spammed with a flood of (bogus) events causing it to constantly close/open the file, potentially blocking log writes as the logger awaits a proper and usable fd.

I think a simple file reload mechanism, as can be seen in many unix daemons, is quite well understood and a cheap solution to this.

**Comment by mholt** [comment]
We generally try to ignore signals since they take on varied and ambiguous purposes here and there, but if it's very important we can try implementing SIGHUP to close and re-open logs. It may result in some lost writes however, unless we sync all writers to block while we re-open the file.

**Comment by motiejus** [comment]
> We generally try to ignore signals since they take on varied and ambiguous purposes here and there, but if it's very important we can try implementing SIGHUP to close and re-open logs. It may result in some lost writes however, unless we sync all writers to block while we re-open the file.

If one does `(*os.File).Write(...)` on Linux, it will be atomic: due to *two* locks: one in Go, one in kernel. I was surprised when I found it (see [this](https://github.com/golang/go/issues/49877) and even more surprised with the resolution.

After reading a bit more about it I found that the problem with locks is *contention*, not having locks per se. If we want to add another lock that waits for the server to re-open the file, contention will barely change, thus throughput.

> I'm a bit busy right now, but could prioritize this for a sufficiently-tiered sponsor; or anyone else is welcome to contribute it!

Thanks for your honest answer! It is not a priority for me right now, but just didn't feel right to leave it without registering an issue first. I am happy we have a consensus. If those lost requests ever become a problem for me (currently I restart caddy after log rotation), I will happily implement it myself.

**Comment by jonathanspw** [comment]
SELinux configs don't really play nicely with caddy and built-in log rotation currently.  I'd prefer to use logrotate anyway for consistency.

It'd be great if a solution here was implemented to allow this.

**Comment by Kovah** [comment]
Hi. Is this a feature that might be implemented soon? It would solve as a compromise for #1096 where time-based log rotation was requested. I would totally be fine to handle rotation via logrotate, but as pointed out in https://github.com/caddyserver/caddy/issues/1096#issuecomment-1758543583 this would require to completely restart caddy after the rotation. If Caddy could adjust the used logs on reload, this would be a good compromise for everybody.
(Sadly, I'm not into Go enough to contribute this on my own.)

**Comment by mholt** [comment]
I personally don't have the time to work on it before 2.9, but pull requests are welcomed!

**Comment by carlwgeorge** [comment]
The RPMs from the caddy copr now set the appropriate selinux context on `/var/log/caddy` (see https://github.com/caddyserver/dist/issues/137), so that the caddy binary can write there with selinux enforcing.  I don't know if that is enough to resolve the issue described here, but it's worth checking out.

**Comment by IslamElsayed** [comment]
@mholt I'd like to pick this up, since you said PRs were welcome. A few things I found reading the current code, and then the questions I'd want settled before writing anything — because I think one of them is a judgement call rather than an implementation detail.

**Where it stands on master (`c18099a0`).** `sigtrap_posix.go` traps SIGHUP and does nothing with it:

```go
case syscall.SIGHUP:
    // ignore; this signal is sometimes sent outside of the user's control
    Log().Info("not implemented", zap.String("signal", "SIGHUP"))
```

**The part I think is easy to miss.** `FileWriter.OpenWriter` has two quite different paths, and only one of them is actually broken:

```go
roll := fw.Roll == nil || *fw.Roll
...
file, err := os.OpenFile(fw.Filename, os.O_WRONLY|os.O_APPEND|os.O_CREATE, modeIfCreating)
...
if roll {
    _ = file.Close() // timberjack will reopen it on its own
    closeFile = false
}
```

With `roll` at its default of true, the writer is a `*timberjack.Logger`, which owns the file and already has `Rotate()`. The plain `*os.File` is only returned when someone sets `roll: false` — and that is exactly what you configure when you want an external tool like logrotate to own rotation. So the users who hit this bug are precisely the ones who opted out of Caddy's rolling, which makes the gap self-selecting and, I think, argues for fixing it rather than telling people to use `roll`.

On the mechanics: `writers` is a package-level `UsagePool` (`logging.go:806`) and `UsagePool.Range` already exists (`usagepool.go:149`), so enumerating the live writers to reopen them is available without new bookkeeping.

**The questions, most important first.**

1. **Does implementing this conflict with the stance the comment records?** That line says SIGHUP is ignored *because* "this signal is sometimes sent outside of the user's control", and you've said you generally avoid signals since they "take on varied and ambiguous purposes". Acting on SIGHUP means a stray hangup now closes and reopens log files. If that trade still bothers you, I'd rather build it as an explicit opt-in — a config flag, or an admin API endpoint — than quietly change what a signal does. Your call, and it changes the shape of the PR.
2. **The lost-writes problem you raised.** You noted it "may result in some lost writes unless we sync all writers to block while we re-open the file". I'd take a brief write lock during the swap so no line is written to a closed handle. Is a short blocking pause on every log writer acceptable to you, or would you rather have a lock-free swap that may drop lines?
3. **Scope.** Only `roll: false` file writers, which are the genuinely broken case? Or also call `Rotate()` on the timberjack-backed ones, so SIGHUP means "rotate now" uniformly? The first is narrower and fixes the reported problem; the second is more predictable but changes behaviour for the default configuration.
4. You asked in 2023 what happens with more than one log output — my assumption is reopen every file-backed writer and leave stdout/stderr/discard alone, but say if you'd rather it were narrower.

Acceptance would be a test that opens a `FileWriter` with `roll: false`, renames the file out from under it, triggers the reopen, and asserts a fresh file exists at the original path and receives subsequent writes — plus a separate assertion that the roll path invokes `Rotate()`, if (3) lands that way.

Happy to be told this still isn't worth the signal-handling complexity.

Disclosure: I use AI assistance when working on patches, and I check and test everything before submitting.

**Comment by Forza-tng** [comment]
How about having a command line option to trigger reopening of log files? Would make it more platform agnostic.

**Comment by IslamElsayed** [comment]
Good point, and I think it's the stronger design.

`sigtrap_nonposix.go` is, in its entirety:

```go
func trapSignalsPosix() {}
```

so a SIGHUP-only fix would leave Windows with no way to reopen logs at all.

Caddy already has the pattern for this: `caddy reload` is a CLI command that POSTs to `/load` on the admin API (`AdminAPIRequest`, `cmd/commandfuncs.go`). So `caddy reopen-logs` → `POST /reopen-logs` would follow the same shape, work on every platform, and still fit logrotate fine, since `postrotate` runs arbitrary commands.

It also sidesteps the objection recorded in the code itself. SIGHUP is ignored today because it "is sometimes sent outside of the user's control", and an explicit admin command carries none of that ambiguity.

So I'd revise my question 1 above: the admin endpoint plus a CLI wrapper as the real mechanism, and SIGHUP as an optional convenience on POSIX calling the same code — or dropped entirely if signals are unwanted here.

@mholt that's the one I'd still want your call on before building, since it decides the shape rather than the details.
